hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/QuasiClosedStuckUnderReplicationHandler.java

```diff
@@ -0,0 +1,158 @@
+  public static final Logger LOG = LoggerFactory.getLogger(QuasiClosedStuckUnderReplicationHandler.class);
+
+  private final PlacementPolicy placementPolicy;
+  private final ReplicationManager replicationManager;
+  private final long currentContainerSize;
+  private final ReplicationManagerMetrics metrics;
+
+  public QuasiClosedStuckUnderReplicationHandler(final PlacementPolicy placementPolicy,
+      final ConfigurationSource conf, final ReplicationManager replicationManager) {
+    this.placementPolicy = placementPolicy;
+    this.currentContainerSize = (long) conf.getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+        ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.replicationManager = replicationManager;
+    this.metrics = replicationManager.getMetrics();
+  }
+
+  @Override
+  public int processAndSendCommands(Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps,
+      ContainerHealthResult result, int remainingMaintenanceRedundancy) throws IOException {
+    ContainerInfo containerInfo = result.getContainerInfo();
+    LOG.debug("Handling under replicated QuasiClosed Stuck Ratis container {}", containerInfo);
+
+    int pendingAdd = 0;
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        pendingAdd++;
+      }
+    }
+
+    if (pendingAdd > 0) {
+      LOG.debug("Container {} has pending add operations. No more replication will be scheduled until they complete",
+          containerInfo);
+      return 0;
+    }
+
+    QuasiClosedStuckReplicaCount replicaCount =
+        new QuasiClosedStuckReplicaCount(replicas, remainingMaintenanceRedundancy);
+
+    List<QuasiClosedStuckReplicaCount.MisReplicatedOrigin> misReplicatedOrigins
+        = replicaCount.getUnderReplicatedReplicas();
+
+    if (misReplicatedOrigins.isEmpty()) {
+      LOG.debug("Container {} is not under replicated", containerInfo);
+      return 0;
+    }
+
+    // Schedule Replicas for the under replicated origins.
+    int totalRequiredReplicas = 0;
+    int totalCommandsSent = 0;
+    IOException firstException = null;
+    List<ContainerReplicaOp> mutablePendingOps = new ArrayList<>(pendingOps);
+    for (QuasiClosedStuckReplicaCount.MisReplicatedOrigin origin : misReplicatedOrigins) {
+      totalRequiredReplicas += origin.getReplicaDelta();
+      List<DatanodeDetails> targets;
+      try {
+        targets = getTargets(containerInfo, replicas, origin.getReplicaDelta(), mutablePendingOps);
+      } catch (SCMException e) {
+        if (firstException == null) {
+          firstException = e;
+        }
+        LOG.warn("Cannot replicate container {} because no suitable targets were found.", containerInfo, e);
+        continue;
+      }
+
+      List<DatanodeDetails> sourceDatanodes = origin.getSources().stream()
+          .map(ContainerReplica::getDatanodeDetails)
+          .collect(Collectors.toList());
```

Comment:
We could filter out dead nodes here and if that leaves us with zero sources, we should log that. 

A node may already be stale when the replication check ran and this container got queued, and it could then die by the time this code in this thread is run.

Reply:
The replicas are retrieved from SCM just before the under rep handler runs (after it has been dequeued), so the replica states are as current as possible. See `ReplicationManager.processUnderReplicatedContainer()`. I guess they could go dead any time, even momentarily after passing this piece of code. It would be of the order of milli-seconds between getting them from containerManager and this code running.